UCS: Refactoring pointer-array datastructure - removing unused "placeholder" and making allocation faster

alex--m opened the pull request:
## What
This patch is removes the unused "placeholder" feature from the pointer-array data-structure, and makes insertion and iteration faster.

## Why
This was done so that this data-structure would be more efficient, and could be used in data-path. Specifically, this patch was introduced to support UCG - Group-based collective operations.

## How
The basic concept is to replace the free-list with the "next free" index - see more in the inline documentation. This allows it to be used as a ring of buffers - especially when the buffers cannot contain a list-link for some reason.

swx-jenkins3 commented:
Can one of the admins verify this patch?

mellanox-github commented:
<details><summary>Mellanox CI: <b>FAILED</b> on 16 of 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 23-Feb-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/6ca075256df86c76530810997b8d1dbcb467405a/_main) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/dbfe1382e29d4f364310c945d1129ed8cf5080e7/hpc-arm-cavium-jenkins_W0) | :x: FAILURE
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/ebb516e83b20144f214ba97a0e3f15f407bad5a1/hpc-arm-hwi-jenkins_W0) | :x: FAILURE
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/9175767511e6a81f0f55351df0a830e25f86a3b2/hpc-test-node-gpu_W0) | :x: FAILURE
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/e9a323167d08611d5856ced0e0743d1bbef1977c/hpc-test-node-gpu_W1) | :x: FAILURE
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/875f9f043c5b6198306b67fa91db0a8b79a9ccf2/hpc-test-node-gpu_W2) | :x: FAILURE
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/48b6832230ea3f406d914c80285d174df9a8b324/hpc-test-node-gpu_W3) | :x: FAILURE
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/29f63e89f8b2844da65d0e0f9e1032acce707d6d/hpc-test-node-legacy_W0) | :x: FAILURE
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/b72bb2e14303b520d01d780441fa9abf46f9f4ab/hpc-test-node-legacy_W1) | :x: FAILURE
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/ed7ce6ce4ba2c446fb6a4faa952acb096655f527/hpc-test-node-legacy_W2) | :x: FAILURE
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/0379defcbcc08fb658c29ed50ee10318faec91f6/hpc-test-node-legacy_W3) | :x: FAILURE
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/06fa72fda3d17bbb2a79736b2ae7ee4d3163e853/hpc-test-node-new_W0) | :x: FAILURE
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/5fe1ea7d150757f463f8d87a09c8b804995c6a8d/hpc-test-node-new_W1) | :x: FAILURE
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/d3258ccd70dcbae0bba2794f0f6fbb6ce85ad5bb/hpc-test-node-new_W2) | :x: FAILURE
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/3e35ebf76d345f955415fbd70e13203b3b6ca626/hpc-test-node-new_W3) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/1074dddf95d36f3093c3caf16042ad83dd683e84/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/a6f039ccb2d0ca79629869da52fa97627035eba9/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/32a68efcd9fe5ba67712cf3c030868441e0bbad2/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/28147eb3c845180cc4f4e5f8c33ca86cbe0fc465/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/bb91732a6adaabaae896802d827df0a6d1894452/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/a9a7437b503a756d22ec1a72e055a10278ec259c/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/4eb6c6abe50a6fa5d088cbb3aecc3c200d36d43f/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/a7ec63b88b9503bfe5e63ad320f5378e7d38d41c/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/e77b23b8f51dec785fe1bde8ce767747502134e4/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/dfbfbfe892bb3b170053a931e804d1d6/raw/0294ebdf91aacf7f5fd6477655b2a898630f56cf/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

yosefe commented:
seems it makes insert operation O(n) now instead of O(1) ? since freelist is removed

mellanox-github commented:
<details><summary>Mellanox CI: <b>FAILED</b> on 16 of 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 24-Feb-2020

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/1dd43c3070a6680ceba0ca6a2a581757f2c281ca/_main) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/b8a2c53e6e11cc18f4ec38d62d0bbf256aadc6e8/hpc-arm-cavium-jenkins_W0) | :x: FAILURE
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/41a82d42cd9d4fd871cc2353edc21a8b09d45f34/hpc-arm-hwi-jenkins_W0) | :x: FAILURE
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/b7f8293522b917a33dda68620d8ca03b5a4064eb/hpc-test-node-gpu_W0) | :x: FAILURE
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/8b315bbb4e3fc0bbb517e7df513fa6285c28c4d1/hpc-test-node-gpu_W1) | :x: FAILURE
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/2d2cc2de25c3e2918ae75fe7e267e1ba8ef6a9fb/hpc-test-node-gpu_W2) | :x: FAILURE
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/d644e1083b59814157b662a73acb4a50f305b162/hpc-test-node-gpu_W3) | :x: FAILURE
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/b6a0bacb6522ba24b17b210a7ac6adc6e50cddb2/hpc-test-node-legacy_W0) | :x: FAILURE
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/4bc731ca15faf3b5cf7e26d28effac3a0223dd1d/hpc-test-node-legacy_W1) | :x: FAILURE
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/aa50494346452a6866d2ee6570fb7f534e7b394f/hpc-test-node-legacy_W2) | :x: FAILURE
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/c62a79624bcf8e76b8aae5ac1cdc7e3920d39dab/hpc-test-node-legacy_W3) | :x: FAILURE
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/de1be8e794e8c93b126f7acb03800cebf4080aa6/hpc-test-node-new_W0) | :x: FAILURE
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/134d860a302d7ff88df527703b76cdb01173efd5/hpc-test-node-new_W1) | :x: FAILURE
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/a16bbaffd9f932e53f75c5c7e60878fa98da4084/hpc-test-node-new_W2) | :x: FAILURE
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/e8a4ea524389e503b2531e905f65966c12117ccc/hpc-test-node-new_W3) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/809ba8309957070567b3562624d2060f022532de/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/3f1141cb6c3b7ec0484236dc71205a9921c16b13/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/f283501c2eebfe838e05d6f19b504c93533ff19f/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/ebad66954ae2815184473bef42b45f24625c7a2b/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/24d1120061f1e3c06430ed4c0306b59501051ffc/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/5207bc9fb0b4a8bdc7a4d4db7077fe1991247c63/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/39a3fd4915ffc78955151d698e34acd449e919e6/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/87bb871b02d1af57cf33ce02fb4a37e440cc2f8b/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/9b1d34f095ce6aaa70a79ef1b90cee1621602a40/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/977e04c9f1268c4f567be7e469f1fa77/raw/0166027c798870e2fd8bd2f21217fdfa57e58f57/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

alex--m commented:
@yosefe I replaced `freelist` with `first_free`, which works similarly: points at the first free element in the array, so that allocation is _usually_ O(1). Actually, the main difference (for met) is that I replaced the (unused) placeholder with the offset to the next non-free element, to make foreach() faster. I use the pointer array to store descriptors to incoming AM messages I can't process yet (there's a complicated reason why I can't use a member of the descriptor to form a list).

alex--m commented:
I have an engineer in my team working on an improved implementation.

On Wed, 4 Mar 2020, 22:55 Pavel Shamis (Pasha), <notifications@github.com>
wrote:

> @yosefe <https://github.com/yosefe> @alex--m <https://github.com/alex--m>
> Folks, where it stands ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openucx/ucx/pull/4780?email_source=notifications&email_token=AB6EZZZZHDB5ARMDWVHXCXDRF2553A5CNFSM4KV44X5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEN2HHOY#issuecomment-594834363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6EZZZSSTLA77RUT2ZE5I3RF2553ANCNFSM4KV44X5A>
> .
>

alex--m commented:
@shuki-zanyovka - you're up :)

alex--m commented:
Replaced by #4860 .
